lib/charms/tls_certificates_interface/v0/tls_certificates.py

```diff
@@ -0,0 +1,422 @@
+        try:
+            certificate_data[key] = json.loads(raw_relation_data[key])
+        except json.decoder.JSONDecodeError:
+            certificate_data[key] = raw_relation_data[key]
+    return certificate_data
+
+
+class CertificatesProviderCharmEvents(CharmEvents):
+    """List of events that the TLS Certificates provider charm can leverage."""
+
+    certificate_request = EventSource(CertificateRequestEvent)
+
+
+class CertificatesRequirerCharmEvents(CharmEvents):
+    """List of events that the TLS Certificates requirer charm can leverage."""
+
+    certificate_available = EventSource(CertificateAvailableEvent)
+
+
+class InsecureCertificatesProvides(Object):
+    """TLS certificates provider class to be instantiated by TLS certificates providers."""
+
+    on = CertificatesProviderCharmEvents()
+
+    def __init__(self, charm, relationship_name: str):
+        super().__init__(charm, relationship_name)
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+        self.charm = charm
+        self.relationship_name = relationship_name
+
+    @staticmethod
+    def _relation_data_is_valid(certificates_data: dict) -> bool:
+        """Uses JSON schema validator to validate relation data content.
+
+        :param certificates_data: Certificate data dictionary as retrieved from relation data.
+        :return: True/False depending on whether the relation data follows the json schema.
+        """
+        try:
+            validate(instance=certificates_data, schema=REQUIRER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError:
+            return False
+
+    def set_relation_certificate(self, certificate: Cert, relation_id: int):
+        """Adds certificates to relation data.
+
+        :param certificate: Certificate object
+        :param relation_id: Relation ID
+        """
+        logging.info(f"Setting Certificate to {certificate} for {self.model.unit}")
+        certificates_relation = self.model.get_relation(
+            relation_name=self.relationship_name, relation_id=relation_id
+        )
+        relation_data = certificates_relation.data[self.model.unit]  # type: ignore[union-attr]
+
+        current_ca = relation_data.get("ca")
+        current_chain = relation_data.get("chain")
+        if not current_ca:
+            relation_data["ca"] = certificate["ca"]
+        if not current_chain:
+            relation_data["chain"] = certificate["ca"]
+        certificate_dict = {"key": certificate["key"], "cert": certificate["cert"]}
+        relation_data[certificate["common_name"]] = json.dumps(certificate_dict)
+
+        certificate_dict = {"key": certificate["key"], "cert": certificate["cert"]}
+        relation_data[certificate["common_name"]] = json.dumps(certificate_dict)
+
+    def _on_relation_changed(self, event):
+        logger.info(f"Raw relation data: {event.relation.data}")
```

Comment:
```suggestion
        logger.debug("Raw relation data: %s", event.relation.data)
```

Reply:
Consider dropping, but certainly demoting to debug

Reply:
Agreed, fixed